pkg/controllers/controlplanemachineset/controller.go

```diff
@@ -228,9 +228,97 @@ func (r *ControlPlaneMachineSetReconciler) reconcileMachines(ctx context.Context
 // Once the owner references are removed, it removes the finalizer to allow the garbage collector to reap
 // the deleted ControlPlaneMachineSet.
 func (r *ControlPlaneMachineSetReconciler) reconcileDelete(ctx context.Context, logger logr.Logger, cpms *machinev1.ControlPlaneMachineSet) (ctrl.Result, error) {
+	logger.V(1).Info("Reconciling control plane machine set deletion")
+
+	machinesTypeMeta, err := providers.GetMachinesTypeMeta(cpms.Spec.Template.MachineType)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("could not determine machines type: %w", err)
+	}
+
+	logger.V(4).Info("Detected machines type", "kind", machinesTypeMeta.Kind, "version", machinesTypeMeta.APIVersion)
+
+	selector, err := metav1.LabelSelectorAsSelector(&cpms.Spec.Selector)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("could not convert label selector to selector: %w", err)
+	}
+
+	machinesMeta := &metav1.PartialObjectMetadataList{
+		TypeMeta: machinesTypeMeta,
+	}
+	if err := r.List(ctx, machinesMeta, &client.ListOptions{LabelSelector: selector}); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to list machines: %w", err)
+	}
+
+	var errs []error
+
+	for i := range machinesMeta.Items {
+		machineObjectMeta := machinesMeta.Items[i]
+		logger.V(4).Info("Try to remove owner reference", "machine", machineObjectMeta.Name)
```

Comment:
I don't think we need to say try, else you'd expect another message to say we succeeded, let's do this and then log an error if there's an error
```suggestion
		logger.V(4).Info("Removing owner reference", "machine", machineObjectMeta.Name)
```